### PR TITLE
go-github: add utilities for processing payload messages

### PR DESCRIPTION
**Pull request by gmlewis**
Change-Id: I1c80b8eb7816efc9b83e6b8be956430c68809d5e

### CODE REVIEW

File: messages/messages.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	// signatureHeader is the Github header key used to pass the HMAC hexdigest.
+	signatureHeader = "X-Hub-Signature"
+	// deliveryHeader is the Gihub header key used to pass the globally-unique Github Event ID.
+	deliveryHeader = "X-GitHub-Delivery"
+)
+
+// GenSignature generates a signature for the given payload.
+// hashType possible values are "sha1", "sha256", "sha512".
+// If hashType is empty or invalid, "sha1" is used.
+func GenSignature(payload, secretKey []byte, hashType string) string {
```
Comment (willnorris) [inline messages/messages.go]: Do we really need this function?  I can't think of a case where a user of the go-github library should ever need to generate a signature.  I know that you're using it in tests below, but I actually think we should use hardcoded, known signature values for tests.

File: messages/messages.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+// If hashFunc is nil, sha1.New is used.
+func checkMAC(message, messageMAC, key []byte, hashFunc func() hash.Hash) bool {
+	expectedMAC := genMAC(message, key, hashFunc)
+	return hmac.Equal(messageMAC, expectedMAC)
+}
+
+// EventPayload validates an incoming Github Webhook event request
+// and returns the event ID, event type, and the (JSON) payload.
+// secretKey is the GitHub Webhook secret message.
+func EventPayload(r *http.Request, secretKey []byte) (eventID, event string, payload []byte, err error) {
```
Comment (willnorris) [inline messages/messages.go]: I probably wouldn't return eventID here.  It's not relevant for validating the message signature, and can very easily be parsed off the request by the caller if they want it.

File: messages/messages.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+// ValidateSignature validates the signature for the given payload.
+// sig is the GitHub hash signature delivered in the X-Hub-Signature header.
+// payload is the JSON payload sent by GitHub Webhooks.
+// secretKey is the GitHub Webhook secret message.
+// GitHub docs: https://developer.github.com/webhooks/securing/#validating-payloads-from-github
+func ValidateSignature(sig string, payload, secretKey []byte) error {
+	if sig == "" {
+		return errors.New("missing signature")
+	}
+	var sigLen int
```
Comment (willnorris) [inline messages/messages.go]: sigLen is a slightly confusing name, since this isn't actually the length of the signature, it's the length of the prefix.  An alternate option (which might be easier to read, I'm not sure?) would be to:

```go
sigParts := strings.SplitN(sig, "=", 2)
...
```

File: messages/messages.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
+// Copyright 2016 The go-github AUTHORS. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package messages provides functions for validating payloads from GitHub Webhooks.
+// GitHub docs: https://developer.github.com/webhooks/securing/#validating-payloads-from-github
+package messages
```
Comment (willnorris) [inline messages/messages.go]: depending on how much we can simplify the number of exported methods, we might be able to just move this into the main github package.

Comment (gmlewis) [inline messages/messages.go]: SGTM.
I'll move them to `github/messages*.go`.

File: messages/messages.go (hunk 1 above)
Comment (gmlewis) [inline messages/messages.go]: SGTM.  I'll remove it.

File: messages/messages.go (hunk 2 above)
Comment (gmlewis) [inline messages/messages.go]: SGTM.

File: messages/messages.go (hunk 3 above)
Comment (gmlewis) [inline messages/messages.go]: Much nicer.  Thanks.

**Comment by gmlewis** [comment]
Files updated. PTAL.

**Comment by parkr** [comment]
:shipit: would _love_ to use this.

**Comment by parkr** [comment]
@willnorris Any update here? 🙏

File: github/messages.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+const (
+	// sha1Prefix is the prefix used by Github before the HMAC hexdigest.
+	sha1Prefix = "sha1"
+	// sha256Prefix and sha512Prefix are provided for future compatibility.
+	sha256Prefix = "sha256"
+	sha512Prefix = "sha512"
+	// eventHeader is the Github header key used to pass the event type.
+	eventHeader = "X-Github-Event"
```
Comment (parkr) [inline github/messages.go]: I believe this should be `X-GitHub-Event`, with the capital `H` in `GitHub`. 😄

Comment (shurcooL) [inline github/messages.go]: The same change should probably be applied to the comments too; the canonical brand name spelling is "GitHub", not "Github". But if that's a wider problem than just in this PR, it might be acceptable to address it in a different PR.

Comment (gmlewis) [inline github/messages.go]: @parkr - It turns out that canonical headers don't work that way. You can verify this yourself using `curl` or `bat`.
Here is an example response I just got from the GitHub API:

```
X-Github-Request-Id : 68840050:7F91:291081F8:5755B423
X-Github-Media-Type : github.v3
```

I tried searching for the `RFC` that defines this but came up empty.  I'm sure it is out there somewhere.

As for the comments correcting the brand name spelling... yes, that could all be cleaned up by someone.

Comment (gmlewis) [inline github/messages.go]: It turns out that with all the code I removed from the PR, that those header constants were no longer needed anyway, so I removed them.

**Comment by gmlewis** [comment]
@shurcooL  - when you get a chance, could you PTAL?

File: github/messages.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	buf, err := hex.DecodeString(sigParts[1])
+	if err != nil {
+		return nil, nil, fmt.Errorf("error decoding signature %q: %v", signature, err)
+	}
+	return buf, hashFunc, nil
+}
+
+// ValidatedPayload validates an incoming GitHub Webhook event request
+// and returns the (JSON) payload.
+// secretKey is the GitHub Webhook secret message.
```
Comment (shurcooL) [inline github/messages.go]: I don't have a lot of context on how `ValidatedPayload` is meant to be used (perhaps it's a good idea to provide an example). So I have the following question.

> The client must close the response body when finished with it

Looking at the implementation, I see that ValidatedPayload reads the request body. It does not close the request body.

Is the user expected to close the request body after calling `ValidatedPayload`?

Also, `ValidatedPayload` is not documented in a way that makes it clear it will read the request body. Perhaps it can be inferred from "validates the request, returns JSON payload". But maybe it should be explicitly stated that it reads the request body and doesn't close it, if it's indeed the user's responsibility to close it?

Or should it close the request body itself?

File: github/messages_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+			req.Header.Set(signatureHeader, test.signature)
+		}
+
+		got, err := ValidatedPayload(req, secretKey)
+		if err != nil {
+			if test.wantPayload != "" {
+				t.Errorf("ValidatedPayload(%#v): err = %v, want nil", test, err)
+			}
+			continue
+		}
```
Comment (shurcooL) [inline github/messages_test.go]: Related to my comment above, this test code is not valid because it results in the request body not being closed after being read.

**Comment by shurcooL** [comment]
I've spotted one issue with the request body and/or documentation. Aside from that, I'm not seeing any other issues. I think if we resolve that, this will have my LGTM.

File: github/messages.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+func ValidateSignature(signature string, payload, secretKey []byte) error {
+	messageMAC, hashFunc, err := messageMAC(signature)
+	if err != nil {
+		return err
+	}
+	if !checkMAC(payload, messageMAC, secretKey, hashFunc) {
+		return errors.New("payload signature check failed")
+	}
+	return nil
+}
```
Comment (shurcooL) [inline github/messages.go]: Also, what's the difference between `ValidatedPayload` and `ValidateSignature`? Is it intentional that one uses adjective "validated" and the other uses verb "validate"? It seems unintentional, since they are both phrased as "X validates ...":

```
// ValidatedPayload validates ...
// ValidateSignature validates ...
```

The more exported symbols there are, the harder the API is to learn and to use (users will need to learn both, and figure out which one is more appropriate for them to use). Is it beneficial to have both of them exported, or can we make one of them unexported?

`ValidatedPayload` looks to be a simple wrapper on top of `ValidateSignature`.

File: github/messages.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	// signatureHeader is the GitHub header key used to pass the HMAC hexdigest.
+	signatureHeader = "X-Hub-Signature"
+)
+
+// genMAC generates the HMAC signature for a message provided the secret key and hashFunc.
+// If hashFunc is nil, sha1.New is used.
+func genMAC(message, key []byte, hashFunc func() hash.Hash) []byte {
+	if hashFunc == nil {
+		hashFunc = sha1.New
+	}
```
Comment (shurcooL) [inline github/messages.go]: Style, since `checkMAC` is the only caller of `genMAC`, I think `genMAC` can be simplified by factoring the "If hashFunc is nil, sha1.New is used." behavior out to `checkMAC`. I think this is simpler, isn't it?

```Go
// genMAC generates the HMAC signature for a message provided the secret key and hashFunc.
func genMAC(message, key []byte, hashFunc func() hash.Hash) []byte {
	mac := hmac.New(hashFunc, key)
	mac.Write(message)
	return mac.Sum(nil)
}

// checkMAC reports whether messageMAC is a valid HMAC tag for message.
// If hashFunc is nil, sha1.New is used.
func checkMAC(message, messageMAC, key []byte, hashFunc func() hash.Hash) bool {
	if hashFunc == nil {
		hashFunc = sha1.New
	}
	expectedMAC := genMAC(message, key, hashFunc)
	return hmac.Equal(messageMAC, expectedMAC)
}
```

Comment (shurcooL) [inline github/messages.go]: Actually, even `checkMAC` is only called in one place, so you can factor out the default hash func logic even higher, into `ValidateSignature` itself.

That way, neither `genMAC` nor `checkMAC` need to worry about handling nil hash func.

Comment (gmlewis) [inline github/messages.go]: OK, SGTM.  I'll move the nil has func up to `ValidateSignature`.

File: github/messages.go (hunk 6 above)
Comment (gmlewis) [inline github/messages.go]: Sorry, obviously some more comments are needed.  `ValidatePayload` will be used by a server and GitHub will send events to it. A server does not close the response body... only clients do that, and in this case, GitHub is the client.

How about if I add the comment:
```go
// ValidatePayload is called by the server and therefore does not need to close the request body.
```

File: github/messages.go (hunk 8 above)
Comment (gmlewis) [inline github/messages.go]: SGTM. I will stop exporting `validateSignature` and rename the other to `ValidatePayload`.

File: github/messages_test.go (hunk 7 above)
Comment (gmlewis) [inline github/messages_test.go]: I believe this test is OK because we are not going through the `http` package or server testing framework, but just calling the function directly and `bytes.Buffer` has no `Close` function.

File: github/messages.go (hunk 6 above)
Comment (willnorris) [inline github/messages.go]: but here we're talking about a request body, not a response body.  Isn't it the server's job to close that when finished reading?  I'm not sure how strict that needs to be with requests.

File: github/messages.go (hunk 9 above)
Comment (gmlewis) [inline github/messages.go]: It turns out that `messageMAC` always returns a valid `hashFunc` (if there are no errors) so the `nil` check is not needed at all.  :smile:

**Comment by gmlewis** [comment]
@shurcooL - I believe I addressed your comments. PTAL.

File: github/messages.go (hunk 6 above)
Comment (gmlewis) [inline github/messages.go]: In https://godoc.corp.google.com/pkg/net/http it says:

> // For server requests the Request Body is always non-nil
> // but will return EOF immediately when no body is present.
> // The Server will close the request body. The ServeHTTP
> // Handler does not need to.

So I don't think it needs closing.  Do you agree?

Comment (willnorris) [inline github/messages.go]: yep, I agree. That sounds pretty clear.

File: github/messages_test.go (hunk 7 above)
Comment (shurcooL) [inline github/messages_test.go]: Ok, I agree. What I said was wrong, I misunderstood the context in which this is being done. I thought it was an outbound client request being simulated, not an inbound server request.

File: github/messages.go (hunk 6 above)
Comment (shurcooL) [inline github/messages.go]: Ah, I completely missed that this is being done on inbound server requests. Right, this is for webhooks, so GitHub is making the request to user code, and this func is used to validate GitHub's incoming request.

> How about if I add the comment:
>
> ```Go
> // ValidatePayload is called by the server and therefore does not need to close the request body.
> ```

Do you think this comment is helpful and clear? I'm not sure. It almost raises more questions than answers, like "by what server"? It could be the HTTP server that the user is running, or GitHub's server.

If you think it's clear, then feel free to leave it as is.

Otherwise, I think it's completely fine to just remove it. The `net/http` package covers this pretty well, so I don't think it's needed to repeat it. As long as people realize that `ValidatePayload` is used to validate requests coming from GitHub. To that end, if issue #361 is going to be resolved by adding a short example of `ValidatePayload` being used somewhere, then that's even more reason to leave this out. I think seeing `ValidatePayload` used in context of `http.HandleFunc` makes it very clear:

```Go
http.HandleFunc("/hook", func(w http.ResponseWriter, r *http.Request) {
    if !github.VerifyPayload(r, os.Getenv("GH_WEBHOOK_SIGNATURE")) {
        http.Error(w, "invalid signature", http.StatusForbidden)
        return
    }
    // Proces payload ...
})
```

This is really nitpicky so feel free to ignore.

Comment (gmlewis) [inline github/messages.go]: SGTM. I've removed the unclear text and replaced with an example to address #361.

**Comment by gmlewis** [comment]
Cool, thanks for your reviews, @shurcooL - I appreciate it!
Submitting.
